src/rust/sysmon-parser/src/util.rs

```diff
@@ -184,52 +184,156 @@ pub(crate) fn unescape_xml<'a, 'b: 'a>(span: &'a StrSpan<'b>) -> Result<Cow<'b,
+    position: usize,
+) -> Result<(&'a str, Option<StrSpan<'a>>)> {
+    let name = match tokenizer.next() {
+        Some(Ok(xmlparser::Token::Attribute { local, value, .. })) if local.as_str() == "Name" => {
+            value.as_str()
+        }
+        _ => {
+            return Err(Error::ParseEventDataXML {
+                message: "expected XML attribute `Name`".to_string(),
+                position,
+            })
+        }
+    };
+
+    let value = match tokenizer.next() {
+        // <Data Name="X" />
+        Some(Ok(xmlparser::Token::ElementEnd {
+            end: xmlparser::ElementEnd::Empty,
+            ..
+        })) => {
+            // Sysmon will always return Data elements for event fields that do not have values
+            // (such as for RuleName when rule names are not specified in the config), so we'll
+            // interpret this None instead of an empty string because we'll treat it the same.
+            None
+        }
+        // <Data Name="X">
+        Some(Ok(xmlparser::Token::ElementEnd {
+            end: xmlparser::ElementEnd::Open,
+            ..
+        })) => {
+            let text = match tokenizer.next() {
+                // <Data Name="X">Y</Data>
+                Some(Ok(xmlparser::Token::Text { text })) => {
+                    // advance tokenizer to ElementEnd::Close (</Data>)
+                    match tokenizer.next() {
+                        Some(Ok(xmlparser::Token::ElementEnd {
+                            end: xmlparser::ElementEnd::Close(_, tagname),
+                            ..
+                        })) if tagname.as_str() == "Data" => {}
+                        _ => {
+                            return Err(Error::ParseEventDataXML {
+                                message: "expected `</Data>`".to_string(),
+                                position,
+                            })
+                        }
+                    }
+
+                    Some(text)
+                }
+                // <Data Name="X"></Data>
+                Some(Ok(xmlparser::Token::ElementEnd {
+                    end: xmlparser::ElementEnd::Close(_, tagname),
                     ..
-                } => break,
-                _ => {}
-            }
+                })) if tagname.as_str() == "Data" => {
+                    // In XML this is the same as the empty element <Data Name="X" />
+                    None
+                }
+                _ => {
+                    return Err(Error::ParseEventDataXML {
+                        message: "expected Token::Text".to_string(),
+                        position,
+                    })
+                }
+            };
+
+            text
         }
-        match result {
-            Some(r) => r,
-            None => continue,
+        _ => {
+            return Err(Error::ParseEventDataXML {
+                message: "unexpected XML token".to_string(),
+                position,
+            })
```

Comment:
(FWIW it looks like all these messages are always &'static strs; you could change the error signature and save an alloc; but NBD)

Reply:
Hmmm yeah. I definitely would if this were the happy path, I think I just might for this as well.

Reply:
Okay yeah EZ